clfdomainsplit.1
.TH "clfdomainsplit" "1" "0.06" "russell@coker.com.au" "1"
.SH "NAME"
clfdomainsplit \- split Common\-Log Format web logs based on domain name

.SH "SYNOPSIS"
.B clfdomainsplit [\-\-help] [\-i input] [\-d defaultfile] [\-c cfg\-file] [\-o directory]

.SH "DESCRIPTION"
The
.B clfdomainsplit
program will split up large CLF format web logs based on domain name.  This is for
creating separate log analysis passes for each domain hosted on your server.

.SH "OVERVIEW"
The
.BR 
input
parameter specifies the file to read (default is standard input).
.P
The
.BR 
defaultfile
parameter specifies where data goes if it doesn't have a domain (either it has
an IP address for the server or it doesn't have the server\-name \- the URL is
relative to the root of the web server only).  The default will be to print them
on standard error.
.P
The
.BR cfg\-file
parameter is for specifying the rules for determining what is a different
domain name.  For example www.coker.com.au belongs in the same file as
coker.com.au and abc.coker.com.au because domain names ending in .au have three
major components.  The domain names www.workbenelux.nl and workbenelux.nl
belong in the same file because domain names ending in .nl have two major
components (as do .com, and .gov), wheras anything ending in .va belongs to
the same organization.
The rules are of the form
.BR number:pattern
which lists the number of domain parts which are significant (2 for .com and
.nl domains, 3 for .au and .uk domains, and 1 for .va.  The pattern is used
for a simple string comparison, the default will be:
.TP 
2:com
.TP 
2:nl
.TP 
3:au
.TP 
3:uk
.P
If no config file is specified then it will look for
.BR /etc/clfdomainsplit.cfg .
Of course comments start with #.  Also note that
the first match will be used!
.P
The
.BR directory
parameter is to specify the location for the files to be created (default is
the current directory).  I recommend that you use a directory for this and
nothing else as you never know how many files may be created!

.SH "EXIT STATUS"
.B 0
No errors
.P
.B 1
Bad parameters


.SH "AUTHOR"
This program, its manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>.

.SH "SEE ALSO"
.BR clfsplit (1), clfmerge (1)


debian/changelog
logtools (0.13d) unstable; urgency=low

  * Merged changes from NMU and fixed some more GCC 4.3 warnings.

 -- Russell Coker <russell@coker.com.au>  Fri, 06 Jun 2008 15:34:35 +1000

logtools (0.13c+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with GCC 4.3 (Closes: #417376).

 -- Luk Claes <luk@debian.org>  Sun, 09 Mar 2008 17:03:20 +0000

logtools (0.13c) unstable; urgency=low

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 Apr 2003 17:09:00 +1000

logtools (0.13b) unstable; urgency=low

  * Improved documentation that matches how things work in sarge.

 -- Russell Coker <russell@coker.com.au>  Sun,  1 Sep 2002 15:43:00 +0200

logtools (0.13a) unstable; urgency=LOW

  * Made clfdomainsplit support URLs with the port number.

 -- Russell Coker <russell@coker.com.au>  Sun,  2 Dec 2001 23:58:00 +0100

logtools (0.13) unstable; urgency=LOW

  * Made it work with GCC 3.x.
    Closes: #121284

 -- Russell Coker <russell@coker.com.au>  Mon, 26 Nov 2001 21:46:00 +0100

logtools (0.12b) unstable; urgency=LOW

  * Fixed some minor documentation bugs, mainly related to funnel.
    Closes: #112883

 -- Russell Coker <russell@coker.com.au>  Thu, 20 Sep 2001 15:40:02 +0200

logtools (0.12a) unstable; urgency=LOW

  * Fixed header file errors on funnel for hppa.
    Closes: #104730

 -- Russell Coker <russell@coker.com.au>  Sat, 14 Jul 2001 20:04:55 +0200

logtools (0.12) unstable; urgency=LOW

  * clfmerge -d now converts domain names to lower case.

  * clfdomainsplit now converts all domain names to lower case internally
    (for sorting and matching purposes), it still exposes upper-case domains
    on its output.

 -- Russell Coker <russell@coker.com.au>  Wed, 11 Jul 2001 17:09:00 +0200

logtools (0.11) unstable; urgency=LOW

  * Made the "-d" option of clfmerge handle a URL that doesn't contain any '/'
    characters.

 -- Russell Coker <russell@coker.com.au>  Tue, 12 Jun 2001 11:58:49 +0200

logtools (0.10) unstable; urgency=MEDIUM

  * Made clfdomainsplit handle IP addresses and send them to the default file
    (first code to actually use the default file).

  * Fixed clfmerge so that the -d option will not mess with a URL that is
    already fully qualified (checks for "http://" at the start).

 -- Russell Coker <russell@coker.com.au>  Thu, 7 Jun 2001 23:49:55 +0200

logtools (0.08) unstable; urgency=MEDIUM

  * Made clfdomainsplit easier to hack (the base code could be used for
    splitting any text file if 3 small functions are changed).

  * Make clfdomainsplit use multiple processes to solve file handle limits,
    this code isn't yet tested...

 -- Russell Coker <russell@coker.com.au>  Mon, 8 May 2001 13:50:29 +0200

logtools (0.07) unstable; urgency=MEDIUM

  * Updated the description with more content and fixed some formatting issues.
    Closes: #78976

  * Made it clean the config.cache file (and other config data) before
    packaging the source.
    Closes: #96290

  * Fixed a bug where clfmerge would output the last line twice if there was
    only a single log file and buffering was on.

  * Now "domain name mangling" actually works for clfmerge!

  * Added new clfdomainsplit program.

 -- Russell Coker <russell@coker.com.au>  Mon, 7 May 2001 13:52:41 +0200

logtools (0.06) unstable; urgency=low

  * Add option to change buffer size for clfmerge.

  * Also added an option for "domain name mangling" to clfmerge.  This means
    that the web logs are expected to have a domain name preceeding the regular
    CLF format (as suggested in the Apache documentation for mass virtual
    hosting).
    See http://www.apache.org/DELETE_ME/vhosts/mass.html .

  * Fixed a bunch of errors in the man pages.

 -- Russell Coker <russell@coker.com.au>  Wed, 2 May 2001 18:11:24 +0200

logtools (0.05) unstable; urgency=low

  * Fixed many trivial bugs.

 -- Russell Coker <russell@coker.com.au>  Tue, 5 Dec 2000 21:28:54 +0100

logtools (0.04) unstable; urgency=low

  * Fixed heaps of bugs and ported to Solaris.

 -- Russell Coker <russell@coker.com.au>  Tue, 14 Nov 2000 21:39:00 +0100

logtools (0.01) unstable; urgency=low

  * Initial Release.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 Aug 2000 19:53:08 +0200



clfdomainsplit.cpp
#include <stdio.h>

#if (__GNUC__ >= 3)
#include <ext/hash_map>
#else
#include <hash_map>
#endif

#include <unistd.h>
#include <cstring>
#include <vector>
#include <ctype.h>
#include "logtools.h"

using namespace std;
using namespace __gnu_cxx;

// MAX_FDS is the maximum number of files that will be directly written to
// by one process
#define MAX_FDS  900

// NUM_CHILDREN is the number of child processes that each process may have
// Note that each child may also have that many children.  The limit is the
// size of the process table in your kernel...
#define NUM_CHILDREN  100

// Get the "domain" component of the line.  Change this if you want to split
// on other parts of the name (EG split files on the directory names).  Make
// sure that the value assigned to dom is a valid file name.
// Change this if you want to use this program for splitting files other than
// web logs.
bool getDomain(string &dom, PCCHAR buf)
{
  if(!(buf = index(buf, '[')) ) return true;
  if(!(buf = index(buf, ']')) ) return true;
  if(!(buf = index(buf, '"')) ) return true;
  if(!(buf = strstr(buf, "http://")) ) return true;
  buf += 7;

  PCCHAR end = index(buf, '/');
  if(!end) return true;
  PCCHAR tmp_end = index(buf, ':'); //search for ":80/" at the end and ignore it
  if(tmp_end && tmp_end < end)
    end = tmp_end;
  int len = end - buf;
  PCHAR tmp = new char[len];
  memcpy(tmp, buf, len);
  for(int i = 0; i < len; i++)
    tmp[i] = tolower(tmp[i]);
  dom = string(tmp, int(len));
  free(tmp);
  return false;
}

// The data structure for configuration.  If you change the getDomain() and
// canonDom() functions then you'll probably need to change this and the
// parseConfig() function.
typedef struct
{
  string domain;
  int significant_parts;
} CFG_TYPE;
vector<CFG_TYPE> config;

// Get the canonical name
bool canonDom(string &canon, string &dom)
{
  int last = dom.length() - 1;
  canon = dom;
  if(canon[last] == '.')
  {
    canon.erase(last);
    last--; // for names that end in '.'
  }

  if(isdigit(canon[last]))
    return true; // canon == dom for numeric ID, and it's not valid for split

  int significant_parts = 2;
  unsigned int i;
  for(i = 0; i < config.size(); i++)
  {
    size_t pos = dom.rfind(config[i].domain);
    if(int(pos) != -1 && (pos + config[i].domain.length()) == dom.length())
    {
      significant_parts = config[i].significant_parts;
      break;
    }
  }
  for(i = last; i > 0; i--)
  {
    if(dom[i] == '.')
    {
      significant_parts--;
      if(!significant_parts)
      {
        canon.assign(dom, i + 1, dom.length() - i - 1);
        break;
      }
    }
  }
  return false;
}

void usage();

// This function parses the config file.
bool parseConfig(CPCCHAR configFile)
{
  FILE *fp = fopen(configFile, "r");
  if(!fp)
  {
    fprintf(stderr, "Can't open \"%s\".\n", configFile);
    usage();
  }
  char buf[1024];
  CFG_TYPE cfg;
  while(fgets(buf, sizeof(buf), fp))
  {
    buf[sizeof(buf) - 1] = '\0';
    strtok(buf, "\r\n");
    if(strlen(buf) && buf[0] != '#')
    {
      cfg.significant_parts = atoi(buf);
      if(cfg.significant_parts <= 0 || cfg.significant_parts > 10)
      {
        fprintf(stderr, "Error in config line \"%s\".\n", buf);
        return true;
      }
      char *ptr = index(buf, ':');
      ptr++;
      if(*ptr == '\0')
      {
        fprintf(stderr, "Error in config line \"%s\".\n", buf);
        return true;
      }
      cfg.domain = string(ptr);
      config.push_back(cfg);
    }
  }
  fclose(fp);
  return false;
}


FILE **children[NUM_CHILDREN];
int num_fds = 0;
int newTarget = -1; // -1 means handle locally, anything else means the
                    // current child process target
FILE *input = stdin;
PFILE *defFilePtr;
int num_children = 0;



struct eqstr
{
  bool operator()(string s1, string s2) const
  {
    return s1 == s2;
  }
};

struct hash_string
{
  size_t operator()(const string str) const { return hash<const char *>()(str.c_str()); }
};

typedef hash_map<const string, FILE **, hash_string, eqstr> HASH_TYPE;
HASH_TYPE fd_map;

typedef FILE * PFILE;

void usage()
{
  fprintf(stderr, "usage: clfdomainsplit [filenames]\n"
  "This program splits CLF format web logs into multiple files, one for each\n"
  "domain in the input.\n"
  "[-i input] [-d defaultfile] [-c cfg-file] [-o directory]\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

FILE **forkit(string &dom);
bool printLine(string &dom, PCCHAR buf);
void closeit();
bool newdom(string &dom);
FILE **openOutput(string &dom);

int main(int argc, char **argv)
{
  for(int i = 0; i < NUM_CHILDREN; i++)
  {
    children[i] = NULL;
  }
  PCCHAR configFile = "/etc/clfdomainsplit.cfg";
  FILE *defFile = stderr;

  int int_c;
  optind = 0;
  while(-1 != (int_c = getopt(argc, argv, "c:d:i:o:")) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'c':
        configFile = optarg;
      break;
      case 'd':
        defFile = fopen(optarg, "a");
        if(!defFile)
        {
          fprintf(stderr, "Can't open \"%s\".\n", optarg);
          usage();
        }
      break;
      case 'i':
        input = fopen(optarg, "r");
        if(!input)
        {
          fprintf(stderr, "Can't open \"%s\".\n", optarg);
          usage();
        }
      break;
      case 'o':
        if(chdir(optarg))
        {
          fprintf(stderr, "Can't chdir() to \"%s\".\n", optarg);
          usage();
        }
      break;
    }
  }
  if(parseConfig(configFile))
    return ERR_PARAM;

  defFilePtr = new PFILE;
  *defFilePtr = defFile;

  char buf[4096];
  while(fgets(buf, sizeof(buf), input))
  {
    buf[sizeof(buf) - 1] = '\0';
    strtok(buf, "\r\n");
    string dom;
    if(getDomain(dom, buf))
    {
      fprintf(stderr, "bad line:%s\n", buf);
    }
    else
    {
      if(printLine(dom, buf))
      {
        closeit();
        return 1;
      }
    }
  }
  closeit();
  return 0;
}

bool printLine(string &dom, PCCHAR buf)
{
  if(!fd_map[dom])
  {
    if(newdom(dom))
      return true;
  }
  fprintf(*(fd_map[dom]), "%s\n", buf);
  return false;
}

bool newdom(string &dom)
{
  string canon;
  if(canonDom(canon, dom))
  {
    // send this and anything like it to the default output
    // beware - a file of totally random junk could run us out of RAM!!!
    fd_map[canon] = defFilePtr;
    fd_map[dom] = defFilePtr;
    return false;
  }
  if(fd_map[canon])
  {
    fd_map[dom] = fd_map[canon];
    return false;
  }
  FILE **fpp = openOutput(canon);
  if(!fpp)
    return true;
  fd_map[dom] = fpp;
  return false;
}

FILE **openOutput(string &dom)
{
  if(num_fds > MAX_FDS)
  {
    newTarget++;
    if(newTarget > NUM_CHILDREN)
      newTarget = 0;
  }
  if(newTarget != -1)
  {
    if(children[newTarget] == NULL)
    {
      return forkit(dom); // NB forkit() can call us
    }
    else
    {
      fd_map[dom] = children[newTarget];
      return children[newTarget];
    }
  }

  FILE *fp = fopen(dom.c_str(), "a");
  if(!fp)
  {
    fprintf(stderr, "Can't open/create file \"%s\".\n", dom.c_str());
    return NULL;
  }
  num_fds++;
  FILE **val = new PFILE;
  *val = fp;
  fd_map[dom] = val;
  return val;
}

FILE **forkit(string &dom)
{
  int filedes[2], childFiledes[2];
  if(pipe(filedes) || pipe(childFiledes))
  {
    fprintf(stderr, "Can't create pipe!\n");
    return NULL;
  }
  fflush(NULL);
  int rc = fork();
  if(rc == -1)
  {
    fprintf(stderr, "Can't fork!\n");
    return NULL;
  }
  char c;
  if(rc == 0)
  { // child
    closeit();
    num_fds = 0;
    newTarget = -1;
    num_children = 0;
    for(int i = 0; i < NUM_CHILDREN; i++)
    {
      children[i] = NULL;
    }
    close(childFiledes[0]);
    write(childFiledes[1], &c, 1);
    close(childFiledes[1]);
    close(filedes[1]);
    input = fdopen(filedes[0], "r");
    if(!input)
    {
      fprintf(stderr, "Can't fdopen!\n");
      close(filedes[0]);
      exit(1);
    }
    return openOutput(dom);
  }
  else
  { // parent
    num_children++;
    close(filedes[0]);
    close(childFiledes[1]);
    read(childFiledes[0], &c, 1);
    close(childFiledes[0]);
    FILE *fp = fdopen(filedes[1], "w");
    if(!fp)
    {
      fprintf(stderr, "Can't fdopen!\n");
      return NULL;
    }
    FILE **fpp = new PFILE;
    *fpp = fp;
    children[newTarget] = fpp;
    fd_map[dom] = children[newTarget];
    return children[newTarget];
  }
  return NULL; // should never reach here
}

void closeit()
{
  while(!fd_map.empty())
  {
    HASH_TYPE::iterator st = fd_map.begin();
    FILE **fpp = (*st).second;
    if(fpp)
    {
      if(*fpp)
      {
        fclose(*fpp);
      }
      *fpp = NULL;
    }
    fd_map.erase(st);
  }
}
